File: mcp-server/src/valaris_mcp/tools/skills.py
# Copyright (c) 2026 Valaris Studio
# SPDX-License-Identifier: AGPL-3.0-or-later

from __future__ import annotations

import json

from mcp.server.fastmcp import Context

from valaris_mcp.deprecation import deprecated_tool
from valaris_mcp.errors import handle_api_errors
from valaris_mcp.server import AppContext, mcp

MAX_SKILL_FILE_CHARS = 6000
_TRUNCATION_SUFFIX = "… [truncated]"


@mcp.tool()
@handle_api_errors
async def list_skills(
    workspace_slug: str,
    board_id: str | None = None,
    include_archived: bool = False,
    ctx: Context = None,
) -> str:
    """List workspace skills, or a board's effective skill set with board_id.
    Metadata only (incl. each skill's declared toolsets and, for a board, the
    toolsets its loop grant does not cover); get_skill returns the files.

    Args:
        workspace_slug: Workspace slug.
        board_id: Board UUID or slug: return its effective set.
        include_archived: Include archived skills (workspace listing only).
    """
    app: AppContext = ctx.request_context.lifespan_context
    client = app.client
    if board_id is not None:
        path = f"{client.board(workspace_slug, board_id)}/skills"
    else:
        path = f"{client.ws(workspace_slug)}/skills"
        if include_archived:
            path += "?include_archived=true"
    result = await client.get(path)
    skills = result.get("skills", [])
    return json.dumps(
        {
            "skills": skills,
            "count": len(skills),
            "_hint": (
                "Listings are metadata-only. Call get_skill(workspace_slug, slug) "
                "to fetch a skill's files, then write each one verbatim into your "
                "local skills dir to install it."
            ),
        },
        indent=2,
        default=str,
    )


@mcp.tool()
@handle_api_errors
async def get_skill(
    workspace_slug: str,
    slug: str,
    version: int | None = None,
    file_path: str | None = None,
    ctx: Context = None,
) -> str:
    """Fetch a skill version's files verbatim (with its declared toolsets).
    Long files are truncated; re-fetch one whole with file_path.

    Args:
        workspace_slug: Workspace slug.
        slug: Skill slug.
        version: Omit for the latest published; pass explicitly for a draft.
        file_path: Return this one file in full.
    """
    app: AppContext = ctx.request_context.lifespan_context
    client = app.client
    base = f"{client.ws(workspace_slug)}/skills/{slug}"
    if version is None:
        detail = await client.get(base)
        version = detail.get("latest_published_version")
        if version is None:
            return json.dumps(
                {
                    "skill": {
                        key: detail.get(key)
                        for key in (
                            "id",
                            "slug",
                            "name",
                            "description",
                            "latest_published_version",
                            "origin",
                            "toolsets",
                            "lint_warnings",
                        )
                    },
                    "files": [],
                    "_hint": (
                        "This skill has no published version yet. Fetch a draft "
                        "by passing version explicitly."
                    ),
                },
                indent=2,
                default=str,
            )
    result = dict(await client.get(f"{base}/versions/{version}"))
    files = [dict(file) for file in result.get("files", [])]
    if file_path is not None:
        matched = [file for file in files if file.get("path") == file_path]
        if not matched:
            available = ", ".join(file.get("path", "") for file in files)
            result["files"] = []
            result["_hint"] = (
                f"No file '{file_path}' in this version. Available files: {available}"
            )
            return json.dumps(result, indent=2, default=str)
        result["files"] = matched
        result["_hint"] = (
            "Full file content returned. Write it verbatim into your local skills dir."
        )
        return json.dumps(result, indent=2, default=str)
    truncated_count = 0
    for file in files:
        content = file.get("content", "")
        if len(content) > MAX_SKILL_FILE_CHARS:
            file["content"] = content[:MAX_SKILL_FILE_CHARS] + _TRUNCATION_SUFFIX
            truncated_count += 1
    result["files"] = files
    if truncated_count:
        result["_files_truncated"] = truncated_count
        result["_hint"] = (
            f"{truncated_count} file(s) exceeded {MAX_SKILL_FILE_CHARS} chars and "
            "were truncated — re-fetch each one in full with file_path."
        )
    else:
        result["_hint"] = (
            "Write each file verbatim into your local skills dir to install this skill."
        )
    return json.dumps(result, indent=2, default=str)


@mcp.tool()
@handle_api_errors
async def propose_skill(
    workspace_slug: str,
    slug: str,
    files: list[dict],
    name: str | None = None,
    description: str | None = None,
    board_id: str | None = None,
    ctx: Context = None,
) -> str:
    """Propose a new skill or version as a draft pending human approval;
    nothing is served until approved. Idempotent: returns the pending proposal.

    Args:
        workspace_slug: Workspace slug.
        slug: Skill slug to create or version.
        files: [{"path", "content"}], SKILL.md plus support files, verbatim.
        name: Omit to use SKILL.md frontmatter.
        description: Omit to use SKILL.md frontmatter.
        board_id: Originating board.
    """
    app: AppContext = ctx.request_context.lifespan_context
    client = app.client
    body: dict = {"slug": slug, "files": files}
    if name is not None:
        body["name"] = name
    if description is not None:
        body["description"] = description
    if board_id is not None:
        body["board_id"] = board_id
    result = dict(
        await client.post(f"{client.ws(workspace_slug)}/skills/proposals", body)
    )
    result["_hint"] = (
        "Proposal submitted and pending human approval — nothing is published "
        "yet. Do NOT wait or poll for the decision: a human decides on their "
        "own schedule. Continue your work; if approved, the skill reaches "
        "future runs automatically."
    )
    return json.dumps(result, indent=2, default=str)


@mcp.tool()
@handle_api_errors
async def set_skill_binding(
    workspace_slug: str,
    board_id: str,
    skill_slug: str,
    enabled: bool | None = None,
    pinned_version: int | None = None,
    clear_pin: bool = False,
    ctx: Context = None,
) -> str:
    """Upsert a skill binding on a board: enable/disable, optionally pin a
    version. Omitted fields are unchanged (tri-state); new bindings default
    to enabled, unpinned.

    Args:
        workspace_slug: Workspace slug.
        board_id: Board UUID or slug.
        skill_slug: Workspace skill slug.
        enabled: Active on the board.
        pinned_version: Version to pin.
        clear_pin: Remove the pin; exclusive with pinned_version.
    """
    if clear_pin and pinned_version is not None:
        return json.dumps(
            {
                "error": True,
                "message": (
                    "pinned_version and clear_pin are mutually exclusive — "
                    "pass a version to pin, or clear_pin=true to unpin, "
                    "not both."
                ),
            },
            indent=2,
        )
    app: AppContext = ctx.request_context.lifespan_context
    client = app.client
    body: dict = {}
    if enabled is not None:
        body["enabled"] = enabled
    if clear_pin:
        # Explicit null is the REST tri-state's unpin branch; omission means
        # "leave the pin unchanged".
        body["pinned_version"] = None
    elif pinned_version is not None:
        body["pinned_version"] = pinned_version
    result = await client.put(
        f"{client.board(workspace_slug, board_id)}/skills/{skill_slug}", body
    )
    result["_hint"] = (
        "Binding saved. Runners picking up work on this board (e.g. via "
        "next_assignment) now see this skill in the board's effective set."
    )
    return json.dumps(result, indent=2, default=str)


async def _skill_binding_rows(client, workspace_slug: str, board_id: str) -> dict:
    result = await client.get(f"{client.board(workspace_slug, board_id)}/skills/bindings")
    bindings = result.get("bindings", [])
    return {
        "bindings": bindings,
        "count": len(bindings),
        "_hint": (
            "These are raw binding rows — disabled bindings included. For "
            "the board's effective set (what callers actually get), call "
            "list_skills with board_id. Change a row with "
            "set_skill_binding; remove one with remove_skill_binding."
        ),
    }


@mcp.tool()
@handle_api_errors
async def list_skill_bindings_raw(
    workspace_slug: str,
    board_id: str,
    ctx: Context = None,
) -> str:
    """Raw view of a board's skill binding rows, disabled ones included;
    list_skills(board_id) is the effective set callers actually get.

    Args:
        workspace_slug: Workspace slug.
        board_id: Board UUID or slug.
    """
    app: AppContext = ctx.request_context.lifespan_context
    result = await _skill_binding_rows(app.client, workspace_slug, board_id)
    return json.dumps(result, indent=2, default=str)


@deprecated_tool()
@handle_api_errors
async def list_skill_bindings(
    workspace_slug: str,
    board_id: str,
    ctx: Context = None,
) -> str:
    """List a board's raw skill binding rows.

    Args:
        workspace_slug: Workspace slug.
        board_id: Board UUID or slug.
    """
    app: AppContext = ctx.request_context.lifespan_context
    result = await _skill_binding_rows(app.client, workspace_slug, board_id)
    return json.dumps(result, indent=2, default=str)


@mcp.tool()
@handle_api_errors
async def remove_skill_binding(
    workspace_slug: str,
    board_id: str,
    skill_slug: str,
    ctx: Context = None,
) -> str:
    """Delete a board's skill binding row, pin included; disable via
    set_skill_binding to keep it. Idempotent.

    Args:
        workspace_slug: Workspace slug.
        board_id: Board UUID or slug.
        skill_slug: Bound skill slug.
    """
    app: AppContext = ctx.request_context.lifespan_context
    client = app.client
    await client.delete(
        f"{client.board(workspace_slug, board_id)}/skills/{skill_slug}"
    )
    return json.dumps(
        {
            "removed": True,
            "board_id": board_id,
            "skill_slug": skill_slug,
            "_hint": (
                "Binding removed — the skill is unbound and any version pin is "
                "gone. To merely disable while keeping the pin, use "
                "set_skill_binding with enabled=false instead."
            ),
        },
        indent=2,
        default=str,
    )


@mcp.tool()
@handle_api_errors
async def list_skill_catalog(
    workspace_slug: str,
    ctx: Context = None,
) -> str:
    """List the built-in skill catalog: curated bundles (each declaring the
    toolsets it plays in) not yet in the workspace library.

    Args:
        workspace_slug: Workspace slug.
    """
    app: AppContext = ctx.request_context.lifespan_context
    client = app.client
    result = await client.get(f"{client.ws(workspace_slug)}/skill-catalog")
    entries = result.get("entries", [])
    return json.dumps(
        {
            "entries": entries,
            "count": len(entries),
            "_hint": (
                "Copy an entry into the workspace library with "
                "activate_catalog_skill(workspace_slug, catalog_id) — human "
                "sessions only — then bind it to boards with "
                "set_skill_binding."
            ),
        },
        indent=2,
        default=str,
    )


@mcp.tool()
@handle_api_errors
async def activate_catalog_skill(
    workspace_slug: str,
    catalog_id: str,
    ctx: Context = None,
) -> str:
    """Copy a catalog skill into the workspace library as published v1. Human
    sessions only (runner keys 403). Idempotent.

    Args:
        workspace_slug: Workspace slug.
        catalog_id: Catalog entry id, from list_skill_catalog.
    """
    app: AppContext = ctx.request_context.lifespan_context
    client = app.client
    result = dict(
        await client.post(
            f"{client.ws(workspace_slug)}/skill-catalog/{catalog_id}/activate"
        )
    )
    result["_hint"] = (
        "Skill activated into the workspace library as a published version. "
        "Bind it to a board with set_skill_binding so sessions on that board pick it up."
    )
    return json.dumps(result, indent=2, default=str)

File: frontend/src/pages/documentation/mcp-reference/data/skills.ts
// Copyright (c) 2026 Valaris Studio
// SPDX-License-Identifier: AGPL-3.0-or-later

// Intentionally English-only for Wave 2 per phase4-documentation-page.md §2.

import type { ToolDoc } from "./types";

// Author slice — categories: skills. The workspace skill library and its
// per-board bindings; skill files are served verbatim (open SKILL.md standard).
export const SKILLS_TOOL_DOCS: ToolDoc[] = [
  {
    name: "list_skills",
    category: "skills",
    kind: "read",
    description:
      "List the workspace skill library, or a board's effective skill set, with each skill's declared toolsets. Metadata only — file contents live behind get_skill.",
    params: [
      {
        name: "workspace_slug",
        required: true,
        description: "The URL slug identifying the workspace.",
      },
      {
        name: "board_id",
        required: false,
        description:
          "Board UUID or slug (backend-resolved). When given, returns the board's effective skill set instead of the full workspace library.",
      },
      {
        name: "include_archived",
        required: false,
        description:
          "Workspace listings hide archived skills by default; pass true to include them. Ignored when board_id is given.",
      },
    ],
    gotchas: [
      "Listings never inline file contents — install a skill by calling get_skill and writing each returned file verbatim into your local skills dir.",
    ],
    examplePrompt:
      "Using the Backplane MCP, list the skills available on the <board> board in <workspace>.",
    related: ["get_skill", "set_skill_binding"],
  },
  {
    name: "get_skill",
    category: "skills",
    kind: "read",
    description:
      "Fetch a skill version's files verbatim (SKILL.md plus support files) and its declared toolsets. Omitting version resolves the latest published one.",
    params: [
      {
        name: "workspace_slug",
        required: true,
        description: "The URL slug identifying the workspace.",
      },
      {
        name: "slug",
        required: true,
        description: "The skill slug.",
      },
      {
        name: "version",
        required: false,
        description:
          "Explicit version number. Omit to fetch the latest published version; pass explicitly to fetch a specific version or a draft.",
      },
      {
        name: "file_path",
        required: false,
        description:
          "Path of a single file to return in full, bypassing the per-file truncation cap.",
      },
    ],
    gotchas: [
      "Files longer than 6000 chars are truncated in the multi-file response (flagged via _files_truncated) — re-fetch each one in full with file_path.",
      "A skill with no published version returns empty files; pass version explicitly to read a draft.",
    ],
    examplePrompt:
      "Using the Backplane MCP, fetch the pdf-tools skill from <workspace> and install its files locally.",
    related: ["list_skills", "set_skill_binding"],
  },
  {
    name: "propose_skill",
    category: "skills",
    kind: "write",
    description:
      "Propose a new skill (or version) for the workspace library. Lands as a draft pending human approval; idempotent — re-proposing returns the existing pending proposal.",
    params: [
      {
        name: "workspace_slug",
        required: true,
        description: "The URL slug identifying the workspace.",
      },
      {
        name: "slug",
        required: true,
        description: "The skill slug the proposal creates or versions.",
      },
      {
        name: "files",
        required: true,
        description:
          'The skill\'s files as [{"path": ..., "content": ...}] — SKILL.md plus any support files, contents sent verbatim.',
      },
      {
        name: "name",
        required: false,
        description:
          "Display name. Omit to let the backend take it from the SKILL.md frontmatter, which is authoritative.",
      },
      {
        name: "description",
        required: false,
        description:
          "Description. Omit to let the backend take it from the SKILL.md frontmatter, which is authoritative.",
      },
      {
        name: "board_id",
        required: false,
        description:
          "Board UUID or slug (backend-resolved) to associate the proposal with — e.g. the board whose loop produced it.",
      },
    ],
    gotchas: [
      "Nothing is published until a human approves — poll get_approval_status with the returned approval_id before relying on the skill.",
      "File contents are sent verbatim on this write path — the 6000-char truncation cap only applies to reads via get_skill.",
    ],
    examplePrompt:
      "Using the Backplane MCP, propose a pdf-tools skill for <workspace> from the files in my local skills directory.",
    related: ["list_skills", "get_skill", "get_approval_status"],
  },
  {
    name: "set_skill_binding",
    category: "skills",
    kind: "write",
    description:
      "Enable or disable a workspace skill on a board, optionally pinning a version. Idempotent — re-binding updates in place.",
    params: [
      {
        name: "workspace_slug",
        required: true,
        description: "The URL slug identifying the workspace.",
      },
      {
        name: "board_id",
        required: true,
        description: "The board UUID or slug (backend-resolved).",
      },
      {
        name: "skill_slug",
        required: true,
        description: "The slug of the workspace skill to bind.",
      },
      {
        name: "enabled",
        required: false,
        description:
          "Whether the skill is active on the board. Omit to leave an existing binding's state unchanged; a newly created binding defaults to enabled.",
      },
      {
        name: "pinned_version",
        required: false,
        description:
          "Version number to pin the board to. Omit to leave any existing pin unchanged; an unpinned board tracks the latest published version.",
      },
      {
        name: "clear_pin",
        required: false,
        description:
          "Pass true to remove an existing version pin, returning the board to tracking the latest published version. Mutually exclusive with pinned_version.",
      },
    ],
    gotchas: [
      "Omitting pinned_version leaves an existing pin as-is (PUT semantics of an omitted field) — it does not clear the pin; unpin with clear_pin: true.",
      "pinned_version and clear_pin together is an error — pin or unpin, not both.",
    ],
    examplePrompt:
      "Using the Backplane MCP, enable the pdf-tools skill on the <board> board in <workspace>, pinned to version 2.",
    related: ["list_skills", "get_skill", "next_assignment"],
  },
  {
    name: "list_skill_bindings_raw",
    category: "skills",
    kind: "read",
    description:
      "Raw view of a board's skill binding rows, disabled ones included. list_skills(board_id) is the effective set callers actually get.",
    params: [
      {
        name: "workspace_slug",
        required: true,
        description: "The URL slug identifying the workspace.",
      },
      {
        name: "board_id",
        required: true,
        description: "The board UUID or slug (backend-resolved).",
      },
    ],
    gotchas: [
      "Rows are configuration, not what agents get: a disabled binding or one resolving to no published version appears here but never in the board's effective set (list_skills with board_id).",
    ],
    examplePrompt:
      "Using the Backplane MCP, call list_skill_bindings_raw on the <board> board in <workspace> to show every configured skill binding, including disabled ones.",
    related: ["list_skills", "set_skill_binding", "remove_skill_binding"],
  },
  {
    name: "remove_skill_binding",
    category: "skills",
    kind: "write",
    description:
      "Unbind a skill from a board, deleting the binding row (enabled flag and pin included). Idempotent — removing a missing binding still succeeds.",
    params: [
      {
        name: "workspace_slug",
        required: true,
        description: "The URL slug identifying the workspace.",
      },
      {
        name: "board_id",
        required: true,
        description: "The board UUID or slug (backend-resolved).",
      },
      {
        name: "skill_slug",
        required: true,
        description: "The slug of the bound skill to remove.",
      },
    ],
    gotchas: [
      "Unbind destroys the row's configuration — any version pin is lost. To keep the pin but take the skill out of the effective set, disable it instead via set_skill_binding with enabled: false.",
    ],
    examplePrompt:
      "Using the Backplane MCP, unbind the pdf-tools skill from the <board> board in <workspace>.",
    related: ["set_skill_binding", "list_skill_bindings_raw"],
  },
  {
    name: "list_skill_catalog",
    category: "skills",
    kind: "read",
    description:
      "List the built-in skill catalog: platform-curated skills not yet in the workspace library, each with its declared toolsets, ready to activate.",
    params: [
      {
        name: "workspace_slug",
        required: true,
        description: "The URL slug identifying the workspace.",
      },
    ],
    gotchas: [
      "Catalog entries are not workspace skills yet — activate one (activate_catalog_skill) to copy it into the library before it can be bound to boards.",
    ],
    examplePrompt:
      "Using the Backplane MCP, list the built-in skill catalog for <workspace>.",
    related: ["activate_catalog_skill", "list_skills"],
  },
  {
    name: "activate_catalog_skill",
    category: "skills",
    kind: "write",
    description:
      "Copy a catalog entry into the workspace library as a published v1. Idempotent — re-activation returns the existing copy untouched.",
    params: [
      {
        name: "workspace_slug",
        required: true,
        description: "The URL slug identifying the workspace.",
      },
      {
        name: "catalog_id",
        required: true,
        description: "The catalog entry id, from list_skill_catalog.",
      },
    ],
    gotchas: [
      "Human sessions only: the backend rejects runner-bound callers (API keys linked to a runner identity) with a 403 — activation is a curation decision reserved for people.",
    ],
    examplePrompt:
      "Using the Backplane MCP, activate the tdd-discipline catalog skill into <workspace> and bind it to the <board> board.",
    related: ["list_skill_catalog", "set_skill_binding", "list_skills"],
  },
];
